core/src/main/java/org/apache/calcite/util/ToNumberUtils.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.calcite.sql.SqlBasicTypeNameSpec;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlDataTypeSpec;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.util.regex.Pattern;
+
+/**
+ * A <code>ToNumberUtils</code> implementation for the ToNumber method.
+ */
+public class ToNumberUtils {
+
+  private ToNumberUtils() {
+  }
+
+  private static void handleCasting(
+      SqlWriter writer, SqlCall call, int leftPrec, int rightPrec,
+      SqlTypeName sqlTypeName) {
+    call.setOperand(1,
+        new SqlDataTypeSpec(new SqlBasicTypeNameSpec(sqlTypeName, SqlParserPos.ZERO),
+            SqlParserPos.ZERO));
+    SqlNode[] extractNodeOperands = new SqlNode[]{call.operand(0), call.operand(1)};
+    SqlCall extractCallCast = new SqlBasicCall(SqlStdOperatorTable.CAST,
+        extractNodeOperands, SqlParserPos.ZERO);
+
+    SqlStdOperatorTable.CAST.unparse(writer, extractCallCast, leftPrec, rightPrec);
+  }
+
+  private static void handleNegativeValue(SqlCall call, String regEx) {
+    String firstOperand = call.operand(0).toString().replaceAll(regEx, "");
+    if (call.operand(1).toString().contains("MI") || call.operand(1).toString().contains("S")) {
+      firstOperand = "-" + firstOperand;
+    }
+    call.setOperand(0,
+        new SqlBasicCall(SqlStdOperatorTable.LITERAL_CHAIN, new SqlNode[]{
+            SqlLiteral.createCharString(firstOperand.trim(),
+                call.operand(1).getParserPosition())}, SqlParserPos.ZERO));
+  }
+
+  public static void handleToNumber(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    switch (call.getOperandList().size()) {
+    case 2:
+      if (Pattern.matches("^'[Xx]+'", call.operand(1).toString())) {
+        SqlNode[] sqlNodes = new SqlNode[]{new SqlBasicCall(SqlStdOperatorTable.LITERAL_CHAIN,
+            new SqlNode[]{SqlLiteral.createCharString("0x", call.operand(1).getParserPosition())},
+            SqlParserPos.ZERO), call.operand(0)};
+        SqlCall extractCall = new SqlBasicCall(SqlStdOperatorTable.CONCAT, sqlNodes,
+            SqlParserPos.ZERO);
+        call.setOperand(0, extractCall);
+
+        ToNumberUtils.handleCasting(writer, call, leftPrec, rightPrec, SqlTypeName.INTEGER);
+
+      } else if (call.operand(0).toString().contains(".")) {
+        String regEx = "[-']+";
+        handleNegativeValue(call, regEx);
+        handleCasting(writer, call, leftPrec, rightPrec, SqlTypeName.FLOAT);
+      } else {
+        String regEx = "[-',$]+";
+        handleNegativeValue(call, regEx);
+
+        handleCasting(writer, call, leftPrec, rightPrec, (call.operand(0).toString
+            ().contains("E")
+            && call.operand(1).toString().contains("E"))
```

Comment:
use a variable to make this casting type decision.

Reply:
Done